From: Stefan Bader <stefan.bader@canonical.com>
Date: Fri, 13 Feb 2015 10:40:18 +0100
Subject: libxl: Fix (boot)loader and emulator issues

For the HVM loader and pygrub we always tried to avoid using absolute paths
because in Debian those binaries are in versioned directories and libxl can
actually handle this and will extend the path.

For the device-model/emulator the same was true. It is a bit more complex
now that the libxl driver tries to figure out whether the provided qemu is
the upstream or the "traditional" Xen one.
On the other hand Debian and Ubuntu now both do not ship the traditional
qemu-dm anymore and even before the encouraged setup was to use the upstream
version (for both KVM and Xen).
Since that upstream qemu is in a non-versioned standard path, it is possible
to move to requiring the full pathname here.
The part of fixing up the qemu-dm string (which was a balant lie anyway) is
something upstream probably does not care about.
FIXME: I should rework this after testing to drop any previous removal of
the full path of the emulator string.

Signed-off-by: Stefan Bader <stefan.bader@canonical.com>

Index: libvirt-1.2.12/docs/schemas/domaincommon.rng
===================================================================
--- libvirt-1.2.12.orig/docs/schemas/domaincommon.rng	2015-01-23 12:46:24.000000000 +0100
+++ libvirt-1.2.12/docs/schemas/domaincommon.rng	2015-02-13 10:00:43.161661835 +0100
@@ -258,7 +258,7 @@
                 </choice>
               </attribute>
             </optional>
-            <ref name="absFilePath"/>
+            <ref name="filePath"/>
           </element>
         </optional>
         <optional>
@@ -1060,7 +1060,7 @@
       <optional>
         <element name="bootloader">
           <choice>
-            <ref name="absFilePath"/>
+            <ref name="filePath"/>
             <empty/>
           </choice>
         </element>
Index: libvirt-1.2.12/src/libxl/libxl_conf.c
===================================================================
--- libvirt-1.2.12.orig/src/libxl/libxl_conf.c	2015-02-10 11:30:57.000000000 +0100
+++ libvirt-1.2.12/src/libxl/libxl_conf.c	2015-02-13 10:30:21.709634414 +0100
@@ -428,7 +428,7 @@ libxlCapsInitGuests(libxl_ctx *ctx, virC
         if ((guest = virCapabilitiesAddGuest(caps,
                                              guest_archs[i].hvm ? "hvm" : "xen",
                                              guest_archs[i].arch,
-                                              "qemu-dm",
+                                              "/usr/bin/qemu-system-i386",
                                              (guest_archs[i].hvm ?
                                               "hvmloader" : NULL),
                                              1,
Index: libvirt-1.2.12/src/libxl/libxl_domain.c
===================================================================
--- libvirt-1.2.12.orig/src/libxl/libxl_domain.c	2015-01-23 12:46:24.000000000 +0100
+++ libvirt-1.2.12/src/libxl/libxl_domain.c	2015-02-13 10:39:32.281625925 +0100
@@ -534,6 +534,11 @@ libxlDomainDeviceDefPostParse(virDomainD
         }
     }
 
+    if (def->emulator && STREQ(def->emulator, "qemu-dm")) {
+        VIR_FREE(def->emulator);
+        if (VIR_STRDUP(def->emulator, "/usr/bin/qemu-system-i386") < 0)
+            return -1;
+    }
     return 0;
 }
 
